Sources/PCMultipartAttachmentUploader.swift

```diff
@@ -0,0 +1,174 @@
+import Foundation
+import PusherPlatform
+
+struct PCMultipartAttachmentUploadTask {
+    let uploadRequest: PCMultipartAttachmentUploadRequest
+    let roomID: String
+    let file: Data
+    let partNumber: Int
+}
+
+struct PCMultipartAttachmentUploadResult {
+    let attachmentID: String
+    let partNumber: Int
+    let payload: [String: Any]
+}
+
+class PCMultipartAttachmentUploader {
+    fileprivate let instance: Instance
+    fileprivate let uploadTasks: Array<PCMultipartAttachmentUploadTask>
+    fileprivate let dispatchQueue = DispatchQueue(label: "com.pusher.chatkit.multipart-upload-\(UUID().uuidString)")
+    fileprivate let dispatchGroup = DispatchGroup()
+    fileprivate let uploadSuccesses = PCSynchronizedArray<PCMultipartAttachmentUploadResult>()
+    fileprivate let uploadFailures = PCSynchronizedArray<Error>()
+
+    init(
+        instance: Instance,
+        uploadTasks: [PCMultipartAttachmentUploadTask]
+    ) {
+        self.instance = instance
+        self.uploadTasks = uploadTasks
+    }
+
+    func upload(completionHandler: @escaping ([PCMultipartAttachmentUploadResult]?, [Error]?) -> Void) {
+        for task in self.uploadTasks {
+            self.dispatchGroup.enter()
```

Comment:
You could use PCProgressCounter here? Not sure it simplifies things, so up to you.

Reply:
Hmmm, I didn't realise this existed :D But I think I'm going to leave it be for now.